Initial Webex (Cloud) Integration Support

stevepartridge opened the pull request:
- Support 1:1 Meeting with Webex Cloud 

Things to note:

#### The `/meeting` path
Due to limitations in the Webex cloud API the approach of using a  [`.../plugin/webex/meeting/<meeting ID>`](https://github.com/stevepartridge/mattermost-plugin-webex/blob/add-initial-integration/server/api.go#L21)  was necessary for a couple reasons.  
1) I was unable to find a reliable method of retrieving a meeting URI (nor create one directly).  

2) There is an [open-source project](https://github.com/webex/react-ciscospark) from webex that appears to encourage direct embedding of their [space widget](https://github.com/webex/react-ciscospark/blob/master/packages/node_modules/%40ciscospark/widget-space/README.md) into a third-party integration.  

They are React components and could be embedded directly into the MM webapp, but this approach would require a broader decision across UI/UX, core developers, stakeholders, etc.  Because of this I opted to simply serve up a [basic, full (new/_blank) window page](https://github.com/stevepartridge/mattermost-plugin-webex/blob/add-initial-integration/server/meeting.go#L113-L136) that would deliver the CDN assets from Cisco/Webex and pass in the auth token and 'to user' info.  

#### Webex Account Connecting, OAuth2
Webex Cloud supports standard OAuth2 for easy consent flow connectivity to the authenticated Mattermost user.  In order to ensure a seamless usage across webapp devices/applications (browser desktop) the access/refresh tokens is stored in the KV plugin datastore.  This allows for the user to login and connect once on one session and have it shared in others.  

#### Meeting invite to user not authenticated to Webex
Webex does not appear to allow unknown emails (or user IDs) to be invited to a 1:1 meeting.  Due to this, in the scenario of when a user clicks on the 'JOIN MEETING' button in the post and sent to the meeting/<meeting ID> page, but have not yet connected their webex account, they are run through the OAuth2 consent flow.  To support a better user experience a 302 with a `redirect_to` is sent to the `.../oauth2/connect` path in order to properly redirect the user back to the original meeting once authenticated.  The redirect_to is saved in the KV store with and expiration to ensure clean up of no longer needed entries. 

#### KV Storage Keys Used
| Key | Usage/Intent | Expires |
| ------------- | :----- | ---: |
| **webex_session_**<_mm user id_> | Holds the OAuth2 token (access, refresh, expiry, etc) | No, should it? |
| **webex_user_**<_mm user id_> | Holds the Webex User info | No | 
| **webex_meeting_**<_uuid_> | Holds the unique meeting info From/To users, channel ID, meeting URL | Yes |
| <_uuid state id_>**_**<_mm user id> | Used for OAuth2 state  | Yes |
| <_uuid state id_>**_redir** | Stores the `redirect_to` (when applicable)  | Yes |

TODO:

- [x] Document Webex Integration Setup
- [x] Document Installation and Configuration
- [x] Disable usage in multi-user channels
- [x] Tests, some tests
- [ ] Tests, much tests

stevepartridge committed:
Initial commit

stevepartridge committed:
Govet clean up

stevepartridge committed:
Initial docs

stevepartridge committed:
Initial docs

stevepartridge committed:
Update README.md

levb commented on server/api_utils.go:
```diff
@@ -0,0 +1,38 @@
+)
+
+// JSONResponse is a helper method to write out a json response with status code
+func JSONResponse(w http.ResponseWriter, v interface{}, statusCode int) {
```
👍

levb commented on server/api_utils.go:
```diff
@@ -0,0 +1,38 @@
+	data := map[string]interface{}{
+		"error":       true,
+		"message":     err.Error(),
+		"status_code": statusCode,
```
Why duplicate the _http_ status_code?

stevepartridge replied:
I've run into issues with front end libraries handling errors differently outside of the 200s. So out of habit I include it in error body responses.  Happy to remove it doesn't seem necessary in this use case.

levb commented on server/api_webex.go:
```diff
@@ -0,0 +1,175 @@
+		return
+	}
+
+	fmt.Printf("\n%+v\n\n", channel)
```
(nit) remove

stevepartridge replied:
👍 thanks

levb commented on server/api_webex.go:
```diff
@@ -0,0 +1,175 @@
+
+	p.API.LogDebug("channel", "channel_id", channel.Id, "name", channel.Name, "session_user", sessionUserID, "type", channel.Type)
+
+	// Not sure how reliable this is...
```
@crspeller is there a kosher way of determining if a channel is a DM, and getting the "other" UserID, better than this?

jespino replied:
Nop, actually is what we do in our code. I would do something like checking the channel type before, to make sure with: `channel.Type == model.CHANNEL_DIRECT`

lieut-data replied:
Indeed, but it's annoying that we don't at least have a helper method in `model` to simplify this...

Question: does webex not support calls among multiple people? Could we simplify this code to just fetch the channel members and support non-direct message channels too?

stevepartridge replied:
@jespino Thanks! I'll update it to use the Type check.  That'll make it easier when this is extended to support multiple people.

@lieut-data Webex does support multiple people but it's handled differently than 1:1 by using more of a room concept.  The flow is different enough that I opted to leave out the functionality on this initial pass.  Since they [webex] seem to expect an integration to be embedded within applications I didn't want to build more around the current basic solution with the html/js page.  In case this integration would be preferred by the UI/UX folks to be a more embedded part of the MM application(s).

levb commented on server/api_webex.go:
```diff
@@ -0,0 +1,175 @@
+	}
+
+	if toUserID == sessionUserID {
+		JSONErrorResponse(w, ErrCreateMeetingFailedOwnChannel, http.StatusBadRequest)
```
We should allow for making meetings with oneself, why not? If I just want to start a meeting in my personal room, I'd Cmd-K to myself and click the WebEx button...

stevepartridge replied:
This is a pre-catch error because the Webex meeting will display that error when launched.

levb commented on server/api.go:
```diff
@@ -0,0 +1,79 @@
+
+	switch {
+	// meeting route with id
+	case strings.Index(path, "/meetings/") == 0:
```
2/5 `strings.HasPrefix` would be more intention-revealing

stevepartridge replied:
Agreed, cleaner as well.

levb reviewed (changes requested):
Steve: thank you for the contribution.

Some initial feedback, I am still reviewing.

stevepartridge committed:
Use HasPrefix instead of Index

stevepartridge committed:
Add more static error variables

stevepartridge committed:
Move meeting struct to more appropriate file

stevepartridge committed:
Remove unnecessary logging

stevepartridge committed:
Use static errors, clean up unnecessary logs

stevepartridge committed:
Remove unnecessary logging

stevepartridge committed:
Some initial happy-path tests

stevepartridge committed:
Move test vars to plugin_test

stevepartridge committed:
Clean up comments

stevepartridge committed:
Add configuration docs and asset

stevepartridge committed:
Update README.md

jwilander commented:
Hi @stevepartridge, mentioning @lieut-data and @jespino to get their help reviewing as @levb and @crspeller are OOF

stevepartridge commented:
Thanks! @jwilander

stevepartridge committed:
Discard unused external dependency

jespino commented on server/plugin.go:
```diff
@@ -0,0 +1,43 @@
+)
+
+const (
+	WebexOAuthSessionKey = "webex_session_"
```
nit: In Mattermost code we normally use snake case all caps for constants (As a plugin not need to follow exactly the same code styling, so feel free to ignore this comment unless other core developer has a stronger opinion).

stevepartridge replied:
Good to know.  Personally I prefer all cap snake case for constants but I got tired of golint saying not to use ALL_CAPS in Go names.  Which eventually broke my habit of using them. :(

lieut-data replied:
FWIW, I think we should /fix/ this in Mattermost code, even at the expense of breaking consistency. It's painful for all the reasons @stevepartridge ran into, and very "un-golike". I support doing it right here, especially given that this is a completely separate repository.

jespino commented on server/webex_store.go:
```diff
@@ -0,0 +1,270 @@
+}
+
+// loadWebexUser loads the webex user info from the KV store if present
+func (p *Plugin) loadWebexUser(userID string) (WebexUserInfo, error) {
```
For this cases we normally return `(*WebexUserInfo, error)`, it is not super important, but is more clear in the code that you are returning `nil, err`, instead `webexUser, err` which is kinda confusing sometimes. Actually in certain situations I can want to return an error and a partial result, but in this cases we only use the `webexUser` if the error is `nil`.

jespino replied:
Probably this is applicable to other functions.

stevepartridge replied:
Good point, I can update it to a reference.

jespino commented on webapp/src/components/root/root.jsx:
```diff
@@ -0,0 +1,210 @@
+
+  }
+
+  constructor(props) {
```
This entire function is not needed.

stevepartridge replied:
Removed

jespino commented on webapp/src/components/root/root.jsx:
```diff
@@ -0,0 +1,210 @@
+const {Tooltip, Popover, OverlayTrigger, Modal} = window.ReactBootstrap
+
+export default class Root extends React.PureComponent {
+  static propTypes = {
```
nit: Not needed to follow, but at mattermost we normally use 4 spaces for javascript. Unless any other core committer have an strong opinion here, feel free to choose follow our styling or not.

lieut-data replied:
@hanzei has an `.editorconfig` ready for the sample plugin that might help here -- recommend merging https://github.com/mattermost/mattermost-plugin-sample/pull/34 as such.

stevepartridge replied:
I added the `.editorconfig` and will try to update all the files to follow the MM standards.

jespino commented on webapp/src/components/root/index.js:
```diff
@@ -0,0 +1,36 @@
+  return {
+    visible: isRootModalVisible(state),
+    authenticated: getAuthenticated(state),
+    state,
```
You are passing there the whole state, this is not how is expected to be used redux. Probably makes more sense to change it for something like `confg: getConfig(state)` and use the visible already passed prop instead of `this.props.state.rootModalVisible` used in the `Root` component.

stevepartridge replied:
@jespino I've updated it to the existing `visible` in the props.  Is this more inline with what you were thinking? (I'm not well versed with redux, please forgive my ignorance)

jespino commented on webapp/src/components/post_type_webex/post_type_webex.jsx:
```diff
@@ -0,0 +1,204 @@
+        isRHS: false,
+    };
+
+    constructor(props) {
```
This function is not needed

stevepartridge replied:
cleaned up 👍

jespino commented on webapp/src/constants/svgs.js:
```diff
@@ -0,0 +1,7 @@
+
+export default {
+    VIDEO_CAMERA: "<svg width='14px' height='10px' viewBox='0 0 16 12' version='1.1' xmlns='http://www.w3.org/2000/svg' xmlns:xlink='http://www.w3.org/1999/xlink'> <g stroke='none' stroke-width='1' fill='inherit' fill-rule='evenodd'> <g transform='translate(-696.000000, -34.000000)' fill-rule='nonzero' fill='inherit'> <g transform='translate(-1.000000, 0.000000)'> <g transform='translate(687.0000 ...
+    VIDEO_CAMERA_3: "<svg width='19px' height='19px' viewBox='0 0 19 19'><path fill='#000000' d='M12,3A9,9 0 0,1 21,12A9,9 0 0,1 12,21A9,9 0 0,1 3,12A9,9 0 0,1 12,3M5.94,8.5C4,11.85 5.15,16.13 8.5,18.06C11.85,20 18.85,7.87 15.5,5.94C12.15,4 7.87,5.15 5.94,8.5Z' /></svg>",
```
You can define this as a react node instead as plain text, something like:
```jsx
VIDEO_CAMERA_3:  (
  <svg
    width='19px'
    height='19px'
    viewBox='0 0 19 19'
  >
    <path fill='#000000' d='M12,3A9,9 0 0,1 21,12A9,9 0 0,1 12,21A9,9 0 0,1 3,12A9,9 0 0,1 12,3M5.94,8.5C4,11.85 5.15,16.13 8.5,18.06C11.85,20 18.85,7.87 15.5,5.94C12.15,4 7.87,5.15 5.94,8.5Z' />
  </svg>
),
```

And this way you avoid to have to use `dangerouslySetInnerHTML`

stevepartridge replied:
Updated both of the constants to be react nodes.  @jespino I borrowed this from the zoom plugin, should I do a PR to update that one as well?

jespino commented on webapp/src/actions/index.js:
```diff
@@ -0,0 +1,165 @@
+// Copyright (c) 2017-present Mattermost, Inc. All Rights Reserved.
```
This files is a bit inconsistent, have arrow functions, regular functions, `async/await` approach and `.then`s, I would prefer to have a consistent implementation.

As a reference, in mattermost, for actions, we normally use regular functions and we prefer `async/await` over `then`s (when is possible).

jespino commented on webapp/src/reducers/index.js:
```diff
@@ -0,0 +1,69 @@
+    case ActionTypes.AUTH_CONNECTED:
+      return action.data;
+    case ActionTypes.AUTH_DISCONNECTED:
+      return true;
```
When `AUTH_DISCONNECTED` `authenticated` is `true`?

stevepartridge replied:
ha! oops, I'll update that.

jespino commented on webapp/src/reducers/index.js:
```diff
@@ -0,0 +1,69 @@
+    }
+}
+
+function disconnected(state = true, action) {
```
Should be disconnect just `!connected`? is needed a separate reducer? if we have different states for connected, can be disconnected a selector, instead of another reducer?, should be the same for `authenticated`? Looks like are all handling the same data, and they are just views of that data.

stevepartridge replied:
Removed `disconnected` and `connected` as they are redundant like you mention.  It's only `authenticated` now.

jespino commented on webapp/src/websocket/index.js:
```diff
@@ -0,0 +1,20 @@
+        }
+
+        if (msg.data.success) {
+          getConnected()(store.dispatch, store.getState);
```
this is better called like:
```js
store.dispatch(getConnected());
store.dispatch(notifyCanStartMeeting());
```

stevepartridge replied:
Updated, thank you for the correction of proper usage. This has been really helpful for me.

jespino commented on server/webex.go:
```diff
@@ -0,0 +1,58 @@
+// references:
+//   https://github.com/jbogarin/go-cisco-webex-teams
+//   https://github.com/jbogarin/go-cisco-webex-teams/blob/master/examples/people/main.go#L15-L18
+func NewWebexClient(token string) (*webexteams.Client, error) {
```
This function will never return an error

stevepartridge replied:
Updated to check for token being empty.  Oversight on my part forgetting to put the check in.

jespino commented on server/webex_store.go:
```diff
@@ -0,0 +1,270 @@
+		}
+
+		// unlikely, but just in case to avoid angry panics
+		if webex == nil {
```
The `NewWebexClient` always return a valid client, never a nil, so is safe to skip this check here.

stevepartridge replied:
Removed

stevepartridge committed:
Move client creation to separate file for testability

Add token check before creating client

stevepartridge committed:
Update WebexOAuthSession and WebexUserInfo to refs

For better alignment with Mattermost code standards
Removed unnecessary check for webex nil

stevepartridge committed:
Add more static errors

stevepartridge committed:
Update WebexOAuthSession and WebexUserInfo to refs

stevepartridge committed:
Remove unnecessary constructors

jespino commented on server/webex_store.go:
```diff
@@ -0,0 +1,267 @@
+			"key", key,
+			"error", appErr,
+		)
+		return webexUser, appErr
```
As soon we have the `*WebexUserInfo` as response, normally you put in this cases `return nil, appErr` which is super clear in the intention, but `return webexUser, appErr` still makes me think you still want to return some kind of "partial result" and an error.

stevepartridge replied:
Agreed, updated.

stevepartridge committed:
Fix return handling to show clear intent

lieut-data commented on README.md:
```diff
@@ -0,0 +1,51 @@
+
+#### Create Webex Integration
+
+Connecting Mattermost users to thier Webex accounts requires [creating an integration](https://developer.webex.com/my-apps/new/integration).  
```
Not doing a holistic review of the text here, but spotted a typo:

```suggestion
Connecting Mattermost users to their Webex accounts requires [creating an integration](https://developer.webex.com/my-apps/new/integration).  
```

lieut-data commented on server/api_utils.go:
```diff
@@ -0,0 +1,38 @@
+	w.Header().Set("Content-Type", "application/json")
+	w.WriteHeader(statusCode)
+	w.Write(data)
+
```
nit: spurious blank line, here and below

lieut-data commented on server/api_test.go:
```diff
@@ -0,0 +1,113 @@
+)
+
+func TestConnected_Fail_NotAuthorizedNoHeaderMMUserID(t *testing.T) {
+
```
nit: spurious blank line (here and below)

lieut-data commented on server/api.go:
```diff
@@ -0,0 +1,79 @@
+func (p *Plugin) ServeHTTP(c *plugin.Context, w http.ResponseWriter, r *http.Request) {
+	config := p.getConfiguration()
+	if err := config.IsValid(); err != nil {
+		http.Error(w, "This plugin is not configured.", http.StatusNotImplemented)
```
I'm not sure this is the right status code: `StatusNotImplemented` generally means that the request method (e.g. `GET` or `POST` is not supported). I'd just suggest `http.StatusInternalServerError`.

stevepartridge replied:
Agreed. Mimicked the github plugin on that one.

lieut-data commented on server/api.go:
```diff
@@ -0,0 +1,79 @@
+
+	path := r.URL.Path
+
+	switch {
```
I'd recommend writing this code to use a https://golang.org/pkg/net/http/#ServeMux -- you can initialize it statically or on init, and then just proxy to its `ServeHTTP` implementation.

stevepartridge replied:
I was curious about this approach (followed other plugins).  Using more of a mux/router based approach feels more idiomatic and DRY for sure.  I opted for [chi](https://github.com/go-chi/chi) so I could use the URL param support.  Additionally, I prefer to having proper method handling (i.e. GET, POST, etc.) and indicating if a method isn't implemented when used improperly.  Is this more inline with what you're recommending?

lieut-data commented on server/api.go:
```diff
@@ -0,0 +1,79 @@
+	}
+
+	user, err := p.loadWebexUser(requestorID)
+	switch {
```
This switch statement /probably/ isn't the idiomatic way to write this. Could we change `loadWebexUser` to return `nil, nil` when there is no error /and/ there is no user? Then we can write it more idiomatically as:
```go
if err != nil {
    // ...
}
if user == nil {
    JSONErrorResponse(w, ErrWebexNotConnected, http.StatusUnauthorized)
    return
}
```

lieut-data commented on server/api.go:
```diff
@@ -0,0 +1,79 @@
+
+// handleConnected checks if the user has connected their webex account
+func (p *Plugin) handleConnected(w http.ResponseWriter, r *http.Request) {
+
```
nit: spurious new lines and start/end of method

lieut-data commented on server/api_webex.go:
```diff
@@ -0,0 +1,171 @@
+
+	user, err := p.loadWebexUser(sessionUserID)
+	if err != nil {
+		JSONResponse(w, err, http.StatusInternalServerError)
```
Is it worthwhile distinguishing an error from another status unauthorized here?

lieut-data commented on server/api_webex.go:
```diff
@@ -0,0 +1,171 @@
+}
+
+func (p *Plugin) handleStartMeeting(w http.ResponseWriter, r *http.Request) {
+
```
nit: spurious lines :)

lieut-data commented on server/api_webex.go:
```diff
@@ -0,0 +1,171 @@
+	}
+
+	_, err := p.loadWebexSession(sessionUserID)
+	switch {
```
Suggest simplifying to the more idiomatic:
```go
if err == ErrWebexSessionNotFound || err == ErrWebexSessionNotFound {
    // ...
} else if err != nil {
    // ...
}
```

I won't repeat this feedback everywhere, but I'd recommend simplifying as such.

lieut-data commented on server/api_webex.go:
```diff
@@ -0,0 +1,171 @@
+			"meeting_topic":     "Webex Meeting",
+			"from_webhook":      "true",
+			"override_username": "Webex",
+			"override_icon_url": "https://smartsolutioncomputer.com/upload-img/Products/Cisco/Webex-logo.png",
```
Can we host this locally? It won't load in some environments.

stevepartridge replied:
I've updated it to use the static asset from the webapp side.

lieut-data commented on server/crypt.go:
```diff
@@ -0,0 +1,79 @@
+)
+
+// Lovingly thieved from the github plugin
+//
```
nit: remove?

stevepartridge replied:
Removed, wanted to give due credit :)

lieut-data commented on server/meeting.go:
```diff
@@ -0,0 +1,150 @@
+  <meta charset="utf8">
+
+  <title>Webex Meeting with ` + meetingWithName + `</title>
+  <link rel="stylesheet" href="https://code.s4d.io/widget-space/production/main.css">
```
Can we host this locally?

stevepartridge replied:
I've updated it to use the static asset from the webapp side.

lieut-data commented on server/meeting.go:
```diff
@@ -0,0 +1,150 @@
+    ` + destinationDataAttributes + `
+    />
+  
+  <script src="https://code.s4d.io/widget-space/production/bundle.js"></script>
```
Can we host this locally?

stevepartridge replied:
I've updated it to use the static asset from the webapp side. 
@lieut-data I did it as an 'external' directory.  Is there a preferred standard for this type of asset? (e.g. third-party, vendor, etc.)

lieut-data reviewed (commented):
A few comments, but not a complete review. Thanks for this plugin, @stevepartridge!

lieut-data committed:
Update README.md

Co-Authored-By: stevepartridge <srpartridge@gmail.com>

stevepartridge committed:
Use internal server error for better clarity

stevepartridge committed:
Clean up spurious lines

stevepartridge committed:
Remove unnecessary action

stevepartridge committed:
Clean up debug usage

stevepartridge committed:
Fix and clean up state/props usage

- Use local assets for images
- Format clean up

stevepartridge committed:
Remove unnecessary reducers

stevepartridge committed:
User store.dispatch properly

stevepartridge committed:
Add support for using and loading local assets

stevepartridge committed:
Add .editorconfig to align with plugin sample

See https://github.com/mattermost/mattermost-plugin-sample/pull/34

stevepartridge committed:
Check channel type to bail if not supported

stevepartridge committed:
Add ErrReplacer helper

stevepartridge committed:
Use static local files for meeting assets

stevepartridge committed:
Nil check on person

stevepartridge committed:
Initiate webexUser properly, return nil accordingly

stevepartridge committed:
Remove unnecessary reducers

stevepartridge committed:
Update svg constants to be react nodes

stevepartridge committed:
Remove use of dangerouslySetInnerHTML

stevepartridge committed:
Update to use local logo asset

stevepartridge committed:
Remove comment reference of origin

stevepartridge committed:
Clean up spurious lines

stevepartridge committed:
Update to use a proper ServeMux with routing

stevepartridge committed:
Update tests to changed results

stevepartridge committed:
Fix custom name to webex

stevepartridge committed:
Format updates for eslint

stevepartridge committed:
Format updates for eslint

stevepartridge committed:
Update to use Webex Teams logo in Join Meeting button

stevepartridge committed:
Format updates for eslint

stevepartridge committed:
Update gopkg with latest deps

stevepartridge committed:
Add unit_test tag to exclude/include accordingly

stevepartridge committed:
Add web client test for unit_test

stevepartridge committed:
Format updates for eslint

stevepartridge committed:
Clean up to align with common MM standards

stevepartridge committed:
Clean up, eslint, unused function removal

stevepartridge committed:
ESLint clean up

stevepartridge committed:
Package updates

stevepartridge committed:
Add coverage output to go test

stevepartridge committed:
ESLint clean up

stevepartridge committed:
Fix restraint to version

stevepartridge committed:
Remove switch case usage

stevepartridge committed:
Fix Webex user fall through error

stevepartridge committed:
Move meeting HTML to template

stevepartridge committed:
Move Webex connected to more appropriate file

stevepartridge committed:
Fix desktop cookie passing in fetch

stevepartridge committed:
Fix eslint double quotes

jespino commented on server/api.go:
```diff
@@ -0,0 +1,42 @@
+	}
+
+	if p.mux == nil {
+		p.prepareRoutes()
```
You are already doing this during "OnActivate", so you can assume that this is properly done. and you don't need it here.

jespino commented on server/api_test.go:
```diff
@@ -0,0 +1,110 @@
+	baseConnectedRequest = httptest.NewRequest("GET", "/connected", nil)
+)
+
+func TestConnected_Fail_NotAuthorizedNoHeaderMMUserID(t *testing.T) {
```
We normally use only CamelCasing without any snake casing in the tests names. And probably here would make sense something like:
```go
func TestConnected(t *testing.T) {
    ...
    t.Run("Fail with not authorized no header MMUserID", func(t *testing.T) {
        ...
    })
    t.Run("Fail with Webex Not Connected", func(t *testing.T) {
        ...
    })
    t.Run("Success with Webex Connected", func(t *testing.T) {
        ...
    })
}
```

jespino commented on server/webex_store.go:
```diff
@@ -0,0 +1,257 @@
+			session, err := p.loadWebexSession(userID)
+			if err != nil {
+				p.API.LogError("error looking up session", "error", err.Error())
+				return webexUser, err
```
I guess it should be `nil, err`

jespino commented on server/webex_store.go:
```diff
@@ -0,0 +1,257 @@
+			webex, err := NewWebexClient(session.Token.AccessToken)
+			if err != nil {
+				p.API.LogError("Error creating new webex client", "error", err.Error())
+				return webexUser, err
```
same here

jespino commented on server/webex_store.go:
```diff
@@ -0,0 +1,257 @@
+			person, _, err := webex.People.GetMe() // don't need resp, supressing it
+			if err != nil {
+				p.API.LogError("Error calling people.GetMe()", "error", err.Error())
+				return webexUser, err
```
same here

jespino commented on server/webex_store.go:
```diff
@@ -0,0 +1,257 @@
+			err = p.storeWebexUser(userID, webexUser)
+			if err != nil {
+				p.API.LogError("Error saving webex user", "error", err.Error())
+				return webexUser, err
```
here I'm not sure if we want to return `nil, err` or `webexUser, err` because in this case we can still use the webexUser. The question is, are we using it if we have an error?

jespino replied:
And the answer is "no", we only use the webexUser when `err == nil` so I would suggest or return here `nil, err` or handle this cases in the `handleOAuthCallback` handler. I think I prefer the former but I'm not 5/5 here.

jespino commented on webapp/src/components/post_type_webex/post_type_webex.jsx:
```diff
@@ -0,0 +1,220 @@
+                    <i
+                        style={style.buttonIcon}
+                    >
+                        <svg
```
makes sense to move it to the Svgs module? maybe we replace the current data structure and export functions components that renders the Svgs, and maybe in this cases accepts a "sytle" prop. It is just a possible improvement, If you don't see it or see it over-complication, feel free to discard this comment.

jespino commented on webapp/src/components/root/root.jsx:
```diff
@@ -0,0 +1,227 @@
+                    <Modal.Header
+                        closeButton={true}
+                        style={style.header}
+                    ></Modal.Header>
```
This should be just a self-closed tag.
```suggestion
                   <Modal.Header
                        closeButton={true}
                        style={style.header}
                    />
```

stevepartridge committed:
Remove unnecessary nil check

stevepartridge committed:
More tests and clean up existing

stevepartridge committed:
Remove unnecessary method check

jespino committed:
Update webapp/src/components/root/root.jsx

Co-Authored-By: stevepartridge <srpartridge@gmail.com>

stevepartridge committed:
Fix syntax

stevepartridge committed:
Update README.md

stevepartridge committed:
Clean up HTML

stevepartridge committed:
Fix comment name
